docs/source/links/state.md

```diff
@@ -0,0 +1,609 @@
+a fragment to an existing object in the cache. This is useful if you want to add
+some client-side fields to an existing object in the cache.
+
+The `id` should correspond to the object's cache key. If you're using the
+`InMemoryCache` and not overriding the `dataObjectFromId` config property, your
+cache key should be `__typename:id`.
+
+```js
+const user = {
+  Mutation: {
+    updateUserEmail: (_, { id, email }, { cache }) => {
+      const data = { email };
+      cache.writeData({ id: `User:${id}`, data });
+      return data;
+    },
+  },
+};
+```
+
+`cache.writeData` should cover most of your needs; however, there are some cases
+where the data you're writing to the cache depends on the data that's already
+there. In that scenario, you should use the DataProxy methods on the Apollo
+cache, which allow you to pass in a query or a fragment.
+
+<h2 id="cache-api">Cache API</h2>
+
+The [Apollo cache API](docs/react/features/caching.html) has several other
+methods to assist you with updating & retrieving data. Here are some common use
+cases where it's best to use the DataProxy methods instead of `cache.writeData`.
+
+<h3 id="write-query">writeQuery & readQuery</h3>
+
+Sometimes, the data you're writing to the cache depends on data that's already
+in the cache; for example, you're adding an item to a list or setting a property
+based on an existing property value. In that case, you should use
+`cache.readQuery` to pass in a query and read a value from the cache before you
+write any data. Let's look at an example where we add a todo to a list:
+
+```js
+let nextTodoId = 0;
+
+const todos = {
+  Query: {
+    todos: () => [],
+  },
+  Mutation: {
+    addTodo: (_, { text }, { cache }) => {
+      const query = gql`
+        query GetTodos {
+          todos @client {
+            id
+            text
+            completed
+          }
+        }
+      `;
+      const previous = cache.readQuery({ query });
+      const data = {
+        todos: previous.todos.concat([
+          { id: nextTodoId++, text, completed: false, __typename: 'TodoItem' },
+        ]),
+      };
+
+      // you can also do cache.writeData({ data }) here if you prefer
+      cache.writeQuery({ query, data });
+      return data;
+    },
+  },
+};
+```
+
+In order to add our todo to the list, we need the todos that are currently in
+the cache, which is why we call `cache.readQuery` to retrieve them.
+`cache.readQuery` will throw an error if the data isn't in the cache, so we need
+to provide an initial state. This is why we're returning an empty array in our
+`Query.todos` resolver.
+
+To write the data to the cache, you can use either `cache.writeQuery` or
+`cache.writeData`. The only difference between the two is that
+`cache.writeQuery` requires that you pass in a query to validate that the shape
+of the data you're writing to the cache is the same as the shape of the data
+required by the query. Under the hood, `cache.writeData` automatically
```

Comment:
Are there performance implications here — is automatic construction a trivial or complicated process?

Reply:
I think we don't need to mention that it generates a query, cause that sounds expensive. Actually it's super simple.

Reply:
To be honest, that explanation helped it make sense to me.

Although I wonder — maybe instead of making a writeData that's like writeQuery but you don't specify a query, and then having all the examples say "you can do either of these", maybe it just wraps writeQuery with a version of writeQuery where the query argument is optional?  That is probably more confusing than helpful, just spitballing here :)

Reply:
I'm not a fan of having a lot of optional arguments myself, but maybe that would make sense!